sdk/nodejs/errors.ts

```diff
@@ -81,3 +81,16 @@ export function isGrpcError(err: Error): boolean {
     const code = (<any>err).code;
     return code === grpc.status.UNAVAILABLE || code === grpc.status.CANCELLED;
 }
+
+export class BadRequestError extends Error {
+
+    public readonly __pulumiBadRequestError: boolean = true;
+
+    constructor(public code: grpc.status, message: string, public details?: any) {
+	super(message);
+    }
+
+    public static isInstance(obj: any): obj is BadRequestError {
+	return utils.isInstance<BadRequestError>(obj, "__pulumiBadRequestError");
+    }
+}
```

Comment:
I guess this is the crux of the PR -- how do we design this bit. If we think this requires a bigger discussion, we could merge without this, or just add a single class that in theory later classes could extend/act as factories for. Things I think we want to discuss:

* How opinionated should the errors be. E.g. for `BadRequestError` I feel like we could do better than `details: any` and instead define `details: BadRequestDetails` or some arrow of detail objects etc.
* Does the error type not imply the status code, etc.
* In general, if users are expected to throw these in their programs, I feel like we shouldn't mention gRPC at all -- this is a Pulumi SDK error that abstracts that for you.

Reply:
Yeah, see my inline comments above as well.  I'm adding this to ideation today so we can discuss it.